experiments/202006-redux-lithtml-codemirror/index.html

```diff
@@ -0,0 +1,215 @@
+<html>
+<script src="https://unpkg.com/redux@4.0.5/dist/redux.min.js"></script>
+<script src="https://unpkg.com/codemirror@5.54.0/lib/codemirror.js"></script>
+<script type="module">
+
+    import {html, render} from "https://unpkg.com/lit-html@1.2.1?module"
+    // import * as Redux from "https://unpkg.com/redux@4.0.5?module"
+    // import {foo} from "https://unpkg.com/codemirror@5.54.0?module"
+
+
+    console.log(render(html`<b>hello</b> world`, document.body))
+
+    /*
+    Redux is a very small FP-ish MVC library
+
+    Imagine all your dispatch+controller code were in a single function that contains a giant switch statement.
+      All possible actions (e.g. user keyboard input, "save", "exit", "rename", etc) are
+        handled via case statements.
+      ...the logic for each of these actions is in each case statement.
+    But what is this "logic" doing?
+    In Redux, it is simply transforming the Model in some way
+
+      function controller(currentModel, action) {
+        switch (action.type) {
+
+          case "rename":
+            return {...currentModel, name: action.newName}
+
+          case "clear":
+            return {...currentModel, content: ""}
+
+          ...
+        }
+      }
+
+
+    You hand this method to Redux, and internally it's taking actions that happen (say as a result
+    of something a user does with your app), and calling this big switch method...
+
+      const newModel = controller(model, {type:"rename", newName: "foo2"})
+
+    Current model + action go in, new model comes out. Repeat.
+
+
+
+    Nomenclature change: Redux has different terms though. Model = State, Controller = Reducer. Revised example,
+      in which your "function controller(..." is now "function myReducer(...":
+
+      const newState = myReducer(state, {type:"rename", newName: "foo2"})
+
+    Immutability: it's pretty important that the incoming state object not be side-affected, at least if you want to
+    keep your sanity when debugging stuff.  You can simply use object spread (as above). There's also immer,
+    which I like and which the redux folks appear to still recommend https://stackoverflow.com/a/60819732
+
+
+
+    The "V" in MVC.
+
+    Redux has something it calls a store, you create it using your reducer function and an initial state:
+
+    const initialState = {name: "foo"}
+    const store = Redux.createStore(myReducer, initialState)  // i.e. "controller" and initial "model"
+
+      // think: simplest possible "view rendering function"...but just for debugging purposes
+      function nameConsoleRenderer() {
+        console.log(`current name: ${store.getState().name}`)
+      }
+
+      // think: simplest possible html-based "view rendering function"
+      function nameDomRenderer() {
+        document.body.innerHTML = `Name: <br>${store.getState().name}</br>`
+      }
+
+    Then you add your rendering functions to the store:
+
+      store.subscribe(nameConsoleRenderer)
```

Comment:
how does this know which part of the tree you are interested in? Won't this callback happen every time any thing changes?